File: src/main/java/org/panteleyev/pwdmanager/Serializer.java
/*
 Copyright © 2017-2025 Petr Panteleyev <petr@panteleyev.org>
 SPDX-License-Identifier: BSD-2-Clause
 */
package org.panteleyev.pwdmanager;

import org.panteleyev.commons.xml.XMLStreamWriterWrapper;
import org.panteleyev.pwdmanager.model.Card;
import org.panteleyev.pwdmanager.model.Field;
import org.panteleyev.pwdmanager.model.FieldType;
import org.panteleyev.pwdmanager.model.Note;
import org.panteleyev.pwdmanager.model.Picture;
import org.panteleyev.pwdmanager.model.WalletRecord;
import org.w3c.dom.Element;
import org.w3c.dom.NodeList;
import org.xml.sax.SAXException;
import org.xml.sax.SAXParseException;
import org.xml.sax.helpers.DefaultHandler;

import javax.xml.XMLConstants;
import javax.xml.namespace.QName;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.ParserConfigurationException;
import javax.xml.validation.Schema;
import javax.xml.validation.SchemaFactory;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.UUID;

import static org.panteleyev.pwdmanager.Constants.BUILD_INFO_BUNDLE;

public class Serializer {
    private enum CardClass {
        CARD,
        NOTE,
        UNKNOWN;

        public static CardClass of(String value) {
            try {
                return valueOf(value);
            } catch (Exception ex) {
                return UNKNOWN;
            }
        }
    }

    // Attributes
    private static final QName ATTR_VERSION = new QName("version");
    private static final QName ATTR_CLASS = new QName("recordClass");
    private static final QName ATTR_UUID = new QName("uuid");
    private static final QName ATTR_NAME = new QName("name");
    private static final QName ATTR_TYPE = new QName("type");
    private static final QName ATTR_MODIFIED = new QName("modified");
    private static final QName ATTR_VALUE = new QName("value");
    private static final QName ATTR_PICTURE = new QName("picture");
    private static final QName ATTR_FAVORITE = new QName("favorite");
    private static final QName ATTR_ACTIVE = new QName("active");

    // Tags
    private static final QName WALLET = new QName("wallet");
    private static final QName RECORD = new QName("record");
    private static final QName NOTE = new QName("note");
    private static final QName FIELD = new QName("field");
    private static final QName FIELDS = new QName("fields");
    private static final QName RECORDS = new QName("records");

    private static final String SCHEMA_URL = "/xsd/password-manager.xsd";

    private static final Schema SCHEMA;
    private static final DefaultHandler HANDLER = new DefaultHandler() {
        @Override
        public void error(SAXParseException e) throws SAXException {
            throw e;
        }
    };

    static {
        var schemaFactory = SchemaFactory.newInstance(XMLConstants.W3C_XML_SCHEMA_NS_URI);
        try {
            SCHEMA = schemaFactory.newSchema(Serializer.class.getResource(SCHEMA_URL));
        } catch (SAXException e) {
            // There should not be any exceptions so just fail.
            throw new RuntimeException(e);
        }
    }

    public static void serialize(OutputStream out, List<WalletRecord> records) {
        try (var w = XMLStreamWriterWrapper.newInstance(out)) {
            w.document(WALLET, () -> {
                w.attribute(ATTR_VERSION, BUILD_INFO_BUNDLE.getString(ATTR_VERSION.getLocalPart()));

                w.element(RECORDS, () -> {
                    for (var r : records) {
                        serializeRecord(w, r);
                    }
                });
            });
        }
    }

    public static List<WalletRecord> deserialize(InputStream in) throws ParserConfigurationException,
            SAXException, IOException {

        var factory = DocumentBuilderFactory.newInstance();
        factory.setSchema(SCHEMA);
        factory.setValidating(false);

        var docBuilder = factory.newDocumentBuilder();
        docBuilder.setErrorHandler(HANDLER);

        var doc = docBuilder.parse(in);

        var rootElement = doc.getDocumentElement();
        var records = rootElement.getElementsByTagName(RECORD.getLocalPart());
        return deserializeRecords(records);
    }

    private static void serializeRecord(XMLStreamWriterWrapper w, WalletRecord record) {
        w.element(RECORD, () -> {
            w.attributes(Map.of(
                    ATTR_UUID, record.uuid(),
                    ATTR_NAME, record.name(),
                    ATTR_MODIFIED, record.modified(),
                    ATTR_PICTURE, record.picture(),
                    ATTR_FAVORITE, record.favorite(),
                    ATTR_ACTIVE, record.active()
            ));

            switch (record) {
                case Card card -> {
                    w.attribute(ATTR_CLASS, CardClass.CARD);

                    var fields = card.fields();
                    if (!fields.isEmpty()) {
                        w.element(FIELDS, () -> {
                            for (var f : fields) {
                                serializeField(w, f);
                            }
                        });
                    }

                    w.textElement(NOTE, card.note());
                }
                case Note note -> w.attribute(ATTR_CLASS, CardClass.NOTE)
                        .textElement(NOTE, note.note());
            }
        });
    }

    private static void serializeField(XMLStreamWriterWrapper w, Field field) {
        w.element(FIELD, Map.of(
                ATTR_NAME, field.name(),
                ATTR_TYPE, field.type(),
                ATTR_VALUE, field.value()
        ));
    }

    private static List<WalletRecord> deserializeRecords(NodeList records) {
        var list = new ArrayList<WalletRecord>(records.getLength());

        // children
        for (int i = 0; i < records.getLength(); i++) {
            var item = records.item(i);

            if (item instanceof Element element) {
                var recordClass = CardClass.of(element.getAttribute(ATTR_CLASS.getLocalPart()));
                if (recordClass != null) {
                    var record = switch (recordClass) {
                        case CARD -> deserializeCard(element);
                        case NOTE -> deserializeNote(element);
                        case UNKNOWN -> null;
                    };
                    if (record != null) {
                        list.add(record);
                    }
                }
            }
        }

        return list;
    }

    private static Card deserializeCard(Element element) {
        var uuid = readUuidAttribute(element);
        var name = element.getAttribute(ATTR_NAME.getLocalPart());
        var picture = Picture.of(element.getAttribute(ATTR_PICTURE.getLocalPart()));
        var modified = Long.parseLong(element.getAttribute(ATTR_MODIFIED.getLocalPart()));
        var favorite = Boolean.parseBoolean(element.getAttribute(ATTR_FAVORITE.getLocalPart()));
        var active = readActiveAttribute(element);

        // fields
        var fList = element.getElementsByTagName(FIELD.getLocalPart());
        var fields = new ArrayList<Field>(fList.getLength());
        for (int i = 0; i < fList.getLength(); i++) {
            var f = deserializeField((Element) fList.item(i));
            fields.add(f);
        }

        // note
        var note = "";
        var notes = element.getElementsByTagName(NOTE.getLocalPart());
        if (notes.getLength() > 0) {
            var noteElement = (Element) notes.item(0);
            note = noteElement.getTextContent();
        }

        return new Card(uuid, modified, picture, name, fields, note, favorite, active);
    }

    private static Field deserializeField(Element e) {
        var name = e.getAttribute(ATTR_NAME.getLocalPart());
        var type = FieldType.valueOf(e.getAttribute(ATTR_TYPE.getLocalPart()));
        var value = Field.deserializeValue(type, e.getAttribute(ATTR_VALUE.getLocalPart()));

        return new Field(type, name, value);
    }

    private static WalletRecord deserializeNote(Element element) {
        var uuid = readUuidAttribute(element);
        var name = element.getAttribute(ATTR_NAME.getLocalPart());
        var modified = Long.parseLong(element.getAttribute(ATTR_MODIFIED.getLocalPart()));
        var favorite = Boolean.parseBoolean(element.getAttribute(ATTR_FAVORITE.getLocalPart()));
        var active = readActiveAttribute(element);

        // note
        var note = "";
        var notes = element.getElementsByTagName(NOTE.getLocalPart());
        if (notes.getLength() > 0) {
            var noteElement = (Element) notes.item(0);
            note = noteElement.getTextContent();
        } else {
            // for backward compatibility we also read text content if <note> child is missing
            note = element.getTextContent();
        }

        return new Note(uuid, name, note, favorite, active, modified);
    }

    private static boolean readActiveAttribute(Element element) {
        // For backward compatibility missing 'active' attribute means true
        var activeElement = element.getAttribute(ATTR_ACTIVE.getLocalPart());
        return activeElement.isBlank() || Boolean.parseBoolean(activeElement);
    }

    private static UUID readUuidAttribute(Element element) {
        var uuidString = element.getAttribute(ATTR_UUID.getLocalPart());
        return uuidString.isEmpty() ? UUID.randomUUID() : UUID.fromString(uuidString);
    }
}

File: src/main/kotlin/org/panteleyev/pwdmanager/Serializer.kt
/*
 * Copyright (c) 2017, Petr Panteleyev <petr@panteleyev.org>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this
 *    list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

package org.panteleyev.pwdmanager

import javafx.application.Platform
import javafx.scene.control.TreeItem
import org.w3c.dom.Document
import org.w3c.dom.Element
import java.io.InputStream
import java.io.OutputStream
import java.util.ArrayList
import javax.xml.parsers.DocumentBuilderFactory
import javax.xml.transform.TransformerFactory
import javax.xml.transform.dom.DOMSource
import javax.xml.transform.stream.StreamResult

internal object Serializer {
    // Attributes
    private const val ID_ATTR = "id"
    private const val NAME_ATTR = "name"
    private const val TYPE_ATTR = "type"
    private const val MODIFIED_ATTR = "modified"
    private const val VALUE_ATTR = "value"
    private const val PICTURE_ATTR = "picture"
    private const val EXPANDED_ATTR = "expanded"
    private const val TARGET_ID_ATTR = "targetId"

    // Tags
    private const val FIELD = "field"
    private const val FIELDS = FIELD + "s"
    private const val RECORDS = "records"

    private val DOC_FACTORY: DocumentBuilderFactory = DocumentBuilderFactory.newInstance()

    fun serialize(out: OutputStream, rootItem: TreeItem<Record>) {
        val docBuilder = DOC_FACTORY.newDocumentBuilder()

        val doc = docBuilder.newDocument()
        val rootElement = doc.createElement("root")

        doc.appendChild(rootElement)

        val e = serializeTreeItem(doc, rootItem)
        rootElement.appendChild(e)

        val transformerFactory = TransformerFactory.newInstance()
        val transformer = transformerFactory.newTransformer()
        val source = DOMSource(doc)
        val result = StreamResult(out)
        transformer.transform(source, result)
    }

    fun deserialize(source: InputStream): TreeItem<Record>? {
        val docBuilder = DOC_FACTORY.newDocumentBuilder()

        val doc = docBuilder.parse(source)

        val rootElement = doc.documentElement
        val nodes = rootElement.childNodes

        for (i in 0..nodes.length - 1) {
            val n = nodes.item(i)
            if (n is Element && n.tagName == "Category") {
                return deserializeCategory(n)
            }
        }

        return null
    }

    fun serializeTreeItem(doc: Document, treeItem: TreeItem<Record>): Element {
        val r = treeItem.value

        return doc.createElement(treeItem.value.javaClass.simpleName).apply {
            setAttribute(ID_ATTR, r.id)
            setAttribute(NAME_ATTR, r.name)
            setAttribute(TYPE_ATTR, r.type.name)
            setAttribute(MODIFIED_ATTR, java.lang.Long.toString(r.modified))
            setAttribute(PICTURE_ATTR, r.picture.name)

            // Category
            if (r is Category) {
                setAttribute(EXPANDED_ATTR, java.lang.Boolean.toString(treeItem.isExpanded))
            }

            if (r is Link) {
                setAttribute(TARGET_ID_ATTR, r.targetId)
            }

            // Card - serialize fields
            if (r is Card) {
                val fields = r.fields
                if (!fields.isEmpty()) {
                    val fieldsElement = doc.createElement(FIELDS)
                    appendChild(fieldsElement)

                    fields.forEach {
                        val fe = serializeField(doc, it)
                        fieldsElement.appendChild(fe)
                    }
                }

                val note = r.note
                val noteElement = doc.createElement("note")
                noteElement.textContent = note
                appendChild(noteElement)
            }

            if (r is Note) {
                appendChild(doc.createTextNode(r.text))
            }

            // Children if any
            if (!treeItem.children.isEmpty()) {
                val children = doc.createElement(RECORDS)
                appendChild(children)

                treeItem.children.forEach {
                    val e = serializeTreeItem(doc, it)
                    children.appendChild(e)
                }
            }
        }
    }

    fun serializeField(doc: Document, f: Field): Element {
        return doc.createElement(FIELD).apply {
            setAttribute(NAME_ATTR, f.name)
            setAttribute(TYPE_ATTR, f.type.name)
            setAttribute(VALUE_ATTR, f.value)
        }
    }

    fun deserializeField(e: Element): Field {
        val name = e.getAttribute(NAME_ATTR)
        val type = FieldType.valueOf(e.getAttribute(TYPE_ATTR))
        val value = e.getAttribute(VALUE_ATTR)

        return Field(type, name, value)
    }

    fun deserializeNote(element: Element): TreeItem<Record> {
        val id = element.getAttribute(ID_ATTR)
        val name = element.getAttribute(NAME_ATTR)
        val modified = java.lang.Long.valueOf(element.getAttribute(MODIFIED_ATTR))!!
        val text = element.textContent

        return TreeItem(Note(id = id, modified = modified, name = name, text = text))
    }

    fun deserializeLink(element: Element): TreeItem<Record> {
        val id = element.getAttribute(ID_ATTR)
        val targetId = element.getAttribute(TARGET_ID_ATTR)
        return TreeItem(Link(id, targetId))
    }

    fun deserializeCategory(element: Element): TreeItem<Record> {
        val id = element.getAttribute(ID_ATTR)
        val name = element.getAttribute(NAME_ATTR)
        val type = RecordType.valueOf(element.getAttribute(TYPE_ATTR))
        val picture = Picture.valueOf(element.getAttribute(PICTURE_ATTR))
        val modified = java.lang.Long.valueOf(element.getAttribute(MODIFIED_ATTR))!!

        var expandedString = element.getAttribute(EXPANDED_ATTR)
        if (expandedString.isEmpty()) {
            expandedString = "false"
        }
        val expanded = java.lang.Boolean.valueOf(expandedString)!!

        val category = Category(id, modified, name, type, picture, expanded)
        val result = TreeItem<Record>(category)
        result.isExpanded = expanded

        result.expandedProperty().addListener { _, oldValue, newValue ->
            if (oldValue != newValue) {
                Platform.runLater { MainWindowController.mainWindow?.writeDocument() }
            }
        }

        // children
        val childNodes = element.childNodes
        for (i in 0..childNodes.length - 1) {
            val n = childNodes.item(i)

            if (n is Element && n.tagName == RECORDS) {
                val records = n.getChildNodes()
                for (j in 0..records.length - 1) {
                    val r = records.item(j)

                    if (r is Element) {
                        val recordItem: TreeItem<Record> = when (r.tagName) {
                            "Category" -> deserializeCategory(r)
                            "Card" -> deserializeCard(r)
                            "Note" -> deserializeNote(r)
                            "Link" -> deserializeLink(r)
                            else -> throw IllegalStateException("Illegal file format")
                        }

                        result.children.add(recordItem)
                    }
                }

                break
            }
        }

        return result
    }

    fun deserializeCard(element: Element): TreeItem<Record> {
        val id = element.getAttribute(ID_ATTR)
        val name = element.getAttribute(NAME_ATTR)
        val picture = Picture.valueOf(element.getAttribute(PICTURE_ATTR))
        val modified = java.lang.Long.valueOf(element.getAttribute(MODIFIED_ATTR))!!

        // fields
        val fList = element.getElementsByTagName(FIELD)
        val fields = ArrayList<Field>(fList.length)
        for (i in 0..fList.length - 1) {
            val f = deserializeField(fList.item(i) as Element)
            fields.add(f)
        }

        // note
        var note = ""
        val notes = element.getElementsByTagName("note")
        if (notes.length > 0) {
            val noteElement = notes.item(0) as Element
            note = noteElement.textContent
        }

        return TreeItem(Card(id, modified, name, picture, fields, note))
    }
}
